mlir/include/mlir/Dialect/GPU/IR/GPUOps.td

```diff
@@ -1371,11 +1371,26 @@ def GPU_BarrierOp : GPU_Op<"barrier"> {
     accessing the same memory can be avoided by synchronizing work items
     in-between these accesses.
 
+    The address space of visible memory accesses can be modified by adding a
+    list of address spaces required to be visible. By default all address spaces
+    are included.
+
+    ```mlir
+    // only workgroup address spaces accesses required to be visible
+    gpu.barrier memfence [#gpu.address_space<workgroup>]
+    // no memory accesses required to be visible
+    gpu.barrier memfence []
+    // all memory accesses required to be visible
+    gpu.barrier
+    ```
+
     Either none or all work items of a workgroup need to execute this op
     in convergence.
   }];
-  let assemblyFormat = "attr-dict";
+  let assemblyFormat = "(`memfence` $address_spaces^)? attr-dict";
   let hasCanonicalizer = 1;
+  let arguments = (ins OptionalAttr<GPU_AddressSpaceAttrArray> :$address_spaces);
```

Comment:
Nit (assuming you're planning to upstream this): this is usually class inheritance (`Args<>`) or at the top of the block

Reply:
Thanks for your help! Glad you can also get something out of it.
Yes, the intention is to upstream.